crates/bls-crypto/src/bls/batch.rs

```diff
@@ -0,0 +1,99 @@
+use core::panic;
+
+use super::{PublicKey, Signature};
+
+use ark_bls12_377::{Fr, G1Projective};
+use ark_ff::{Field, PrimeField};
+use ark_std::log2;
+
+use crate::{BLSError, HashToCurve};
+
+use rand::RngCore;
+
+#[derive(Default)]
+pub struct Batch {
+    entries: Vec<(PublicKey, Signature)>,
+    message: Vec<u8>,
+    extra_data: Vec<u8>,
+}
+
+const SECURITY_BOUND: usize = 128;
+
+/// Returns a byte count for sizing small exponents up to the maximum size of an Fr field element.
+fn byte_count_from_target_batch_size(size: usize, target_security: usize) -> usize {
+    let target_byte_count = (target_security + (log2(size) as usize) + 7) / 8;
+    let field_byte_count = Fr::size_in_bits() / 8;
+
+    std::cmp::min(target_byte_count, field_byte_count)
+}
+
+impl Batch {
+    /// Constructs a new strict batch verifier context for a given message.
+    pub fn new(message: &[u8], extra_data: &[u8]) -> Batch {
+        Batch {
+            entries: vec![],
+            message: message.to_vec(),
+            extra_data: extra_data.to_vec(),
+        }
+    }
+
+    pub fn add(&mut self, public_key: PublicKey, signature: Signature) {
+        self.entries.push((public_key, signature));
+    }
+
+    pub fn verify<H: HashToCurve<Output = G1Projective>>(
+        &self,
+        hash_to_g1: &H,
+    ) -> Result<(), BLSError> {
+        let mut public_keys = vec![];
+        let mut signatures = vec![];
+
+        let exp_size = byte_count_from_target_batch_size(self.entries.len(), SECURITY_BOUND);
+
+        let exponents = self
```

Comment:
One of the exponents can be 1 at not cost to security. This barely increases efficiency though because even considering very small batches the pairings cost way more.

Reply:
Maybe we can do this later as part of an optimization pass (which I imagine also including the sparse exponent work, etc)